src/rust/engine/process_execution/src/nailgun/mod.rs

```diff
@@ -0,0 +1,254 @@
+        NailgunCommandRunner {
+            inner: Arc::new(runner),
+            nailguns: NailgunPool::new(),
+            metadata: metadata,
+            workdir_base: std::env::temp_dir(),
+            python_distribution_absolute_path: python_distribution_absolute_path,
+        }
+    }
+
+    // Ensure that the workdir for the given nailgun name exists.
+    fn get_nailguns_workdir(&self, nailgun_name: &NailgunProcessName) -> Result<PathBuf, String> {
+        let workdir = self.workdir_base.clone().join(nailgun_name);
+        if self.workdir_base.exists() {
+            debug!("Creating nailgun workdir at {:?}", self.workdir_base);
+            std::fs::create_dir_all(workdir.clone())
+                .map_err(|err| format!("Error creating the nailgun workdir! {}", err))
+                .map(|_| workdir)
+        } else {
+            debug!("nailgun workdir {:?} exits. Reusing that...", self.workdir_base);
+            Ok(workdir)
+        }
+    }
+
+    fn calculate_nailgun_name(main_class: &String, digest: &Digest) -> String {
+        format!("{}_{}", main_class, digest.0)
+    }
+}
+
+impl super::CommandRunner for NailgunCommandRunner {
+    fn run(
+        &self,
+        req: MultiPlatformExecuteProcessRequest,
+        workunit_store: WorkUnitStore) -> BoxFuture<FallibleExecuteProcessResult, String> {
+
+        let mut client_req = self.extract_compatible_request(&req).unwrap();
+
+        if !client_req.is_nailgunnable {
+            trace!("The request is not nailgunnable! Short-circuiting to regular process execution");
+            return self.inner.run(req, workunit_store)
+        }
+        debug!("Running request under nailgun:\n {:#?}", &client_req);
+
+        // Separate argument lists, to form distinct EPRs for (1) starting the nailgun server and (2) running the client in it.
+        let ParsedArgLists {nailgun_args, client_args } = split_args(&client_req.argv);
+        let nailgun_req = get_nailgun_request(nailgun_args, client_req.input_files, client_req.jdk_home.clone());
+        trace!("Extracted nailgun request:\n {:#?}", &nailgun_req);
+
+        let maybe_jdk_home = nailgun_req.jdk_home.clone();
+
+        let main_class = client_args.iter().next().unwrap().clone(); // We assume the last one is the main class name
+        let nailgun_req_digest = crate::digest(MultiPlatformExecuteProcessRequest::from(nailgun_req.clone()), &self.metadata);
+        let nailgun_name = NailgunCommandRunner::calculate_nailgun_name(&main_class, &nailgun_req_digest);
+        let nailgun_name2 = nailgun_name.clone();
+        let nailgun_name3 = nailgun_name.clone();
+
+        let nailguns_workdir = try_future!(self.get_nailguns_workdir(&nailgun_name));
+
+        // Materialize the directory for running the nailgun server, if we need to.
+        let workdir_path2 = nailguns_workdir.clone();
```

Comment:
Ok, got it!

I might have been confused by the fact that this looks like it will end up being the client's workdir, but we're actually potentially launching the server here (there _is_ a comment though!). From an encapsulation perspective, if the server was launched closer to the `NailgunPool`, that might be clearer, because `run` would only be dealing with the client.

I believe that what I said above about nailMain is still accurate though: to support JVMs in general (and to get our options isolation), I think we'll want to head in the direction of multiple servers, and no concurrency on each server.

Reply:
> From an encapsulation perspective, if the server was launched closer to the NailgunPool, that might be clearer, because run would only be dealing with the client.

I agree, and I opened https://github.com/pantsbuild/pants/issues/8489 as a follow-up. I spent some hours trying to make it work, but it was growing into a full rewrite of the pool to be futures-aware, so I would like to take the refactoring out of the scope of this PR, if that's okay.